mathics/core/pattern.py

```diff
@@ -289,27 +289,28 @@ def per_expr(yield_expr, expr_groups, sum=0):
 
                         if expr_groups:
                             expr, count = expr_groups.popitem()
-                            max_per_pattern = count / len(patterns)
+                            max_per_pattern = int(count / len(patterns))
                             for per_pattern in range(max_per_pattern, -1, -1):
-                                for next in per_expr(   # nopep8
-                                    expr_groups, sum + per_pattern):
-                                    yield_expr([expr] * per_pattern + next)
+                                perex = per_expr(   # nopep8
```

Comment:
This change was not needed to make the repro snippet from the bug report to work:

```
SetAttributes[or, Orderless]
or[p, p] /. or[p_, p_] :> p
```

but without this change, mathics was crashing when it couldn't match the rule for `Orderless`, like

```
SetAttributes[or, Orderless]
or[p, p, p] /. or[p_, p_] :> p
```

Reply:
I was mistaken about the or[p, p, p].  Please disregard!

Your solution with get_names is just right. I didn't see it, so bravo!